src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,410 @@
+ *    have themselves as the second template argument, e.g.
+ *  ```
+ *  class DataVector : VectorImpl<double, DataVector> {
+ *  ```
+ *  The second template parameter communicates arithmetic type restrictions to
+ *  the underlying Blaze framework. For example, if `VectorType` is
+ *  `DataVector`, then the underlying architecture will prevent the addition the
+ *  vector type with `ModalVector`. Since `DataVector`s and `ModalVector`s
+ *  represent data in different spaces, we wish to forbid several operations
+ *  between them. This vector-type-tracking through an expression prevents
+ *  accidental mixing of vector types in math expressions.
+ *
+ * \note
+ * - If created with size 0, then  `data()` will return `nullptr`
+ * - If `SPECTRE_DEBUG` or `SPECTRE_NAN_INIT` macros are defined, then the
+ *   `VectorImpl` is default initialized to `signaling_NaN()`. Otherwise, the
+ *   vector is filled with uninitialized memory for performance.
+ * - Using an object of this type without explicit initialization is undefined
+     behavior unless `SPECTRE_DEBUG` or `SPECTRE_NAN_INIT` is defined (see
+     previous point).
+ */
+template <typename T, typename VectorType>
+class VectorImpl
+    : public PointerVector<T, blaze::unaligned, blaze::unpadded,
+                           blaze::defaultTransposeFlag, VectorType> {
+  /// \cond HIDDEN_SYMBOLS
+  // the `static_assert` requires a member function around it, as it then
+  // gains access to the fully defined type it is a member of, and therefore
+  // can test the resulting class.
+  static constexpr void private_asserts() noexcept {
+    static_assert(
+        std::is_nothrow_move_constructible<VectorImpl<T, VectorType>>::value,
+        "Missing move semantics");
+  }
+  /// \endcond
+ public:
+  using value_type = T;
+  using size_type = size_t;
+  using difference_type = std::ptrdiff_t;
+  using BaseType = PointerVector<T, blaze::unaligned, blaze::unpadded,
+                                 blaze::defaultTransposeFlag, VectorType>;
+  static constexpr bool transpose_flag = blaze::defaultTransposeFlag;
+
+  using ElementType = T;
+  using TransposeType = VectorImpl<T, VectorType>;
+  using CompositeType = const VectorImpl<T, VectorType>&;
+  using iterator = typename BaseType::Iterator;
+  using const_iterator = typename BaseType::ConstIterator;
+
+  using BaseType::operator[];
+  using BaseType::begin;
+  using BaseType::cbegin;
+  using BaseType::cend;
+  using BaseType::data;
+  using BaseType::end;
+  using BaseType::size;
+
+  // @{
+  /// Upcast to `BaseType`
+  /// \attention
+  /// upcast should only be used when implementing a derived vector type, not in
+  /// calling code
+  const BaseType& operator~() const noexcept {
+    return static_cast<const BaseType&>(*this);
+  }
+  BaseType& operator~() noexcept { return static_cast<BaseType&>(*this); }
+  // @}
+
+  // @{
+  /// Create with the given size and value. In debug mode, the vector is
```

Comment:
I would not have a common doxygen comment here.  I think two separate ones are appropriate, one "Create with the given size." and describing the default (or lack thereof) initialization, and the other "Create with the given size and value."